README.md
# SaveSystem

independent backup system for Linux (tested on Debian GNU / Linux 8.5 (jessie))

---
### Languages used
SaveSystem uses a single language:
 - Shell (100%)

---
### version
0.1.2

Of course this backup system is open source with a [public repository] [save] on GitHub.

---
### Installation

---
#### Getting the source:
To retrieve the source, type:
```sh
$ cd ~ /
$ git clone https://github.com/RootKitDev/SaveSystem.git SaveSystem
```

---
### Use

---
#### General operation
SaveSystem currently runs on the backup server and the script is designed to be
execute at least once a day.

SaveSystem done with any order of priority, the following safeguards:
 - Monthly (Full: e.g full backup of the system)
 - Hebdomadiare (Full: e.g full backup of user / application data)
 - Weekend (the Incremental Monthly: e.g every Saturday)
 - SQL (Full BDDs: e.g every Sunday)
 - Daily (Incremental the Hebdomadiare: the default action if no other backups have not triggered)

SaveSystem integrates a system of "flags" in ```sh /path/to/my/SaveSystem/Flags ```, which allows backup management.
The Flags are by default "arranged" in ```sh /path/to/my/SaveSystem/Flags/Block ``` so that the SaveSystem not interprete the flags by" mistake ".

The list of Flags:
WARNING: Pennants PS arbitrarily priority.
 - EX-000 ( "Exceptional  Backup": execute a backup Monthly out in the condition by default (every 1st of the month))
 - PS-000 ( "No Backup" (Super "No Backup") prevents any kind of backup (without indicating that there was no backup), to let the SQL backups occur )
 - PS-001 ( "No Monthly / Exceptional  Backup" historizes logs in the "No Backup")
 - PS-002 ( "No Hebdomadiare Backup" historizes logs in the "No Backup")
 - PS-003 ( "No Backup Weekend" historizes logs in the "No Backup")
 - PS-004 ( "No daily backup" historizes logs in the "No Backup")

The automatic use of this system requires crontab (or any other task scheduler)
Here is an example of crontab rule

```sh
# Starting the backup script data (Data_Save.sh) every day at 6 am
00 6 * * * /path/to/my/SaveSystem/Data_Save.sh >> /path/to/my/SaveSystem/Logs.d/Cron.log 2> & 1
```

The current system requires a host of reception, for outsourcing backups.

---
### Git Linked
HMI (Human Machine Interface) web is under developement integrating, displaying the backup logs asked CkeckSum display and volume
The [public repository] [HMI] IHM_SaveSystem

---
### Contribution

Want to contribute? Very good !

Send me your ideas and comments by mail: <rootkit.dev@gmail.com>.

---
### Licence

MIT

** Free Software, Hell Yeah! **

---

[Save]: <https://github.com/RootKitDev/SaveSystem>
[HMI]: <https://github.com/RootKitDev/IHM_SaveSystem>


ChangeLog
commit f7ee5a08e9e1a7fbb67dad274eaf2741d7e7e892
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Wed Oct 26 08:23:23 2016 +0200

    Big Upgrade : V1.2 release. Check ChangeLog
    New :
        - Exchange authentication method when SQL dumps : "--defaults-extra-file"
        
        - Adding a system of "partner" in order to ease the verification of availability of "partners" 
        WARNING: host partners in updating manually in case of change (name and / or IP)
        
        -For display in web HMI (update in progress) if more partners are defined,
        the system takes into consideration the statements of previous actions of (KO state does not replace OK state)
        
        - Establishment of calculating the total size of files in the tar ball 
        (instead of the size of the directory that were entered in the file inclusions)

    InComing :
        - An installation and configuration script is writing yard.
        It will activate, or not, a number of "options", and configure up to cater to every need.

            - The options :
                - Outsourcing backups
                - The declaration of the partners (you can not want to host partners)
                - Enable data backups
                - Enable SQL backups

            - The configuration :
                - Declaration of the desired type of backup (Daily, Weekly, Monthly or Weekend (backup Weekly particular)
                - Setups trigger backups (date and time: e.g 1st Saturday of the month)    
        
        
commit a6df6f909c5e1d6f0c81548afae654d9ce1c419d
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Sep 30 13:01:35 2016 +0200

    Upgrade Export Lib for use multi host backup, Upgrade HMI in process

commit f9418c3322a85f7e66eca59d4971d8668d1f1ba8
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Mon Sep 19 14:49:44 2016 +0200

    Export.sh Update : don't delete taball on KT state

commit f43756bef117b6d9be3223deff6a0ced1b2d11cf
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Mon Sep 12 14:05:07 2016 +0200

    Add two folders add repare TARÃR_CMD

commit 625d7f6cbd87a5f61830626f5522c98c990787cf
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Sun Sep 11 17:23:49 2016 +0200

    TMP-Resolv Bug 02 and add begin Revolv incompatibility with IHM_SaveSystem

commit ad786a9fc7d2662fd0056e309fde1c52dae89ecd
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Sun Sep 11 17:23:30 2016 +0200

    TMP-Resolv Bug 02 and add begin Revolv incompatibility with IHM_SaveSystem

commit a5dcd01ea6c81c4aafd8720261810980ecb4a3f5
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Sun Sep 11 17:23:21 2016 +0200

    TMP-Resolv Bug 02 and add begin Revolv incompatibility with IHM_SaveSystem

commit 595edf4c6c0e8c1a030e7e8317fbee5500a69d96
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Thu Sep 8 15:33:20 2016 +0200

    Resolv Bug 01

commit 637d3b5c866102411158f1249235db512574d32f
Merge: 8342a7a c348409
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Sep 2 08:35:59 2016 +0200

    Merge branch 'master' of https://github.com/RootKitDev/SaveSystem

commit 8342a7a23e4730f006bc5a6a2e1e67010ec66944
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Sep 2 08:32:24 2016 +0200

    Add install.sh, Var.sh and config.cfg

commit c3484091cf34b8921d32d3d55f6cdd1af4f3a5a3
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Sep 2 07:46:17 2016 +0200

    Update README.md

commit a32d2bd7818bde164a9937461e9448d01f0d54f8
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Sep 2 07:45:33 2016 +0200

    Update README.md

commit 32dbac4e89eb7ad17db413f6f3f3bb38f2835a98
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Tue Aug 30 14:44:22 2016 +0200

    Update LISEZMOI.md

commit 2cab0fba4f5c0ab2d53dbcae52f0d8f8dc6ab8cc
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Tue Aug 30 14:43:32 2016 +0200

    Update README.md

commit 153f372c934514302890075f7b9f46a3006558aa
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Tue Aug 30 14:41:03 2016 +0200

    Add License, ReadMe, Lisezmoi, ChangeLog file

commit 8c156976fc07d2f83ac2ef95c2c983bba2495179
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Tue Aug 30 09:43:25 2016 +0200

    SQL Save system updated

commit 788632d9d8e8a5f743d3d5d321b54d1b07e837ce
Author: RootKitDev <rootkit.dev@gmail.com>
Date:   Fri Aug 19 14:23:05 2016 +0200

    Initial commit


Data_Save.sh
#! /bin/bash

######################################
# Data_Save.sh
# Utilité: ce script sert à faire des sauvegardés de données de façon dynamique
# Auteur: RootKitDev <RootKit.Dev@gmail.com>
# Mise à jour le: 26/10/2016
######################################

# Verbose mode
for param in $@
do
        if [ $param = "-v" ]
        then
    		set -vx            
        fi
done

# Define PATH

HOME_PATH=""

LIB_PATH="$HOME_PATH/Lib.d"

SUB_LOG=""

# Load Var lib, for all var path (log, flag, ...)
source $LIB_PATH/Var.sh

# Load all the others .sh lib
source $LIB_PATH/exist.sh
source $LIB_PATH/list.sh
source $LIB_PATH/Save.sh
source $LIB_PATH/Export.sh
source $LIB_PATH/required.sh
source $LIB_PATH/Check_Sum.sh
source $LIB_PATH/States.sh

# Writing in the early log
echo "" >> $LOG_PATH/Save.log
echo "-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-" >> $LOG_PATH/Save.log
echo `date` >> $LOG_PATH/Save.log
echo "" >> $LOG_PATH/Save.log

# Determinéetion of useful variables
Now=$(date -d 'now' +"%d-%m-%Y")
Yesterday=$(date -d 'yesterday' +"%d-%m-%Y")

MonthYesterday=$(echo $Yesterday | cut -d'-' -f2)
MonthNow=$(echo $Now | cut -d'-' -f2)

Day=$(date +"%u" | cut -d' ' -f2)

Go_Save_WE="2130"

MonthSave=$(date +"%m_%B")
DaySave=$(date +"%d")

echo "EC" > $SAVE_STATE_PATH/$MonthSave/$DaySave

# Determining the type of backup
if [ $MonthNow -gt $MonthYesterday ] || [ -f $FLAG_PATH/EX-000 ];
	then
		if [ ! -f $FLAG_PATH/PS-001 ] || [ -f $FLAG_PATH/EX-000 ] ;
		then
			# Writing Log File
			if [ -f $FLAG_PATH/EX-000 ];
			then
					echo 'Sauvegarde Exceptionnel (sauvegarde Full) :' >> $LOG_PATH/Save.log
			else
					echo 'Sauvegarde Mensuel (sauvegarde Full) :' >> $LOG_PATH/Save.log
			fi
			
			echo 'Les Répertoires sauvegardés sont :' >> $LOG_PATH/Save.log
		
			if [[ -z "$(echo -e "$(cat $SAVE_LIST_PATH/ListSaveMen)" | tr '\n' ' ' | cut -d ' ' -f1)" ]]; then
				echo -e "\tIl n'y a pas de dossier(s) a sauvegardé(s)" >> $LOG_PATH/Save.log
			else
				
				# Check if directory exist
				for elem in $(cat $SAVE_LIST_PATH/ListSaveMen)
				do
					dir_exist $elem
				done
				
				echo "" >> $LOG_PATH/Save.log
				echo "Eligibilité des dossiers terminée avec succes !" >> $LOG_PATH/Save.log

				# Launch monthly Data Save 
				Data_save "Mensuel"
	
			fi
			
		else
			echo 'Fanion "Pas de Sauvegarde Mensuel" détecté' >> $LOG_PATH/Save.log
			State_Save 4
		fi
elif [[ -e $FLAG_PATH/PS-000 ]];
then

	echo "" >> $LOG_PATH/Save.log
	echo "Fanion \"PS-000\" a été posé," >> $LOG_PATH/Save.log
	echo "aucune sauvegarde de data ne sera lancée" >> $LOG_PATH/Save.log
	State_Save 4

else
		
	if [[ "$Day" = "1" ]];
	then 
		if [ ! -f $FLAG_PATH/PS-002 ];
		then
			# Writing Log File
			echo 'Sauvegarde Hebdomadaire (sauvegarde Full) :' >> $LOG_PATH/Save.log
			echo "" >> $LOG_PATH/Save.log
			echo 'Les Répertoires sauvegardés sont :' >> $LOG_PATH/Save.log
			
			if [[ -z "$(echo -e "$(cat $SAVE_LIST_PATH/ListSaveHeb)" | tr '\n' ' ' | cut -d ' ' -f1)" ]]; then
				echo "Il n'y a pas de dossier(s) a sauvegardé(s)" >> $LOG_PATH/Save.log
			else
				
				# Check if directory exist
				for elem in $(cat $SAVE_LIST_PATH/ListSaveHeb)
				do
					dir_exist $elem
				done
				
				echo "" >> $LOG_PATH/Save.log
				echo "Eligibilité des dossiers terminée avec succes !" >> $LOG_PATH/Save.log

				# Launch weekly Data Save 
				Data_save "Hebdomadaire"
	
			fi
			
		else
			echo 'Fanion "Pas de Sauvegarde Hebdomadaire" détecté' >> $LOG_PATH/Save.log
			State_Save 4
		fi
		
	elif [[ "$Day" = "6" ]] && [[ "$(date +%H%M)" -gt $Go_Save_WE ]];
	then 
		if [ ! -f $FLAG_PATH/PS-003 ];
		then
			# Writing Log File
			echo 'Sauvegarde Week-End (sauvegarde Incrementiel => Mensuel) :' >> $LOG_PATH/Save.log
			echo 'Les Répertoires sauvegardés sont :' >> $LOG_PATH/Save.log
			
			if [[ -z "$(echo -e "$(cat $SAVE_LIST_PATH/ListSaveMen)" | tr '\n' ' ' | cut -d ' ' -f1)" ]]; then
				echo "Il n'y a pas de dossier(s) a sauvegardé(s)" >> $LOG_PATH/Save.log
			else
				
				# Check if directory exist
				for elem in $(cat $SAVE_LIST_PATH/ListSaveMen)
				do
					dir_exist $elem
				done
				
				echo "" >> $LOG_PATH/Save.log
				echo "Eligibilité des dossiers terminée avec succes !" >> $LOG_PATH/Save.log
				
				# Launch weekly-end Data Save 
				Data_save "Week-End"
			fi
			
		else
			echo 'Fanion "Pas de Sauvegarde Week-End" détecté' >> $LOG_PATH/Save.log
			State_Save 4
		fi
	
	elif [ ! -f $FLAG_PATH/PS-004 ];
	then
		# Writing Log File
		echo 'Sauvegarde Journaliere (sauvegarde Incrementiel => Hebdomadaire) :'  >> $LOG_PATH/Save.log
		echo 'Les Répertoires sauvegardés sont :'  >> $LOG_PATH/Save.log
		
		if [[ -z "$(echo -e "$(cat $SAVE_LIST_PATH/ListSaveHeb)" | tr '\n' ' ' | cut -d ' ' -f1)" ]]; then
			echo "Il n'y a pas de dossier(s) a sauvegardé(s)" >> $LOG_PATH/Save.log
		else
			
			# Check if directory exist
			for elem in $(cat $SAVE_LIST_PATH/ListSaveHeb)
			do
				dir_exist $elem
			done
			
			echo "" >> $LOG_PATH/Save.log
			echo "Eligibilité des dossiers terminée avec succes !" >> $LOG_PATH/Save.log

			# Launch daily Data Save 
			Data_save "Journaliere"
		fi
	
	else
		echo 'Fanion "Pas de Sauvegarde Journaliere" détecté' >> $LOG_PATH/Save.log
		State_Save 4
	fi
fi

echo "" >> $LOG_PATH/Save.log
echo "Fin du script $0" >> $LOG_PATH/Save.log
echo `date` >> $LOG_PATH/Save.log